db/migrate/20211029070915_add_treatment_group_membership_constraint.rb

```diff
@@ -0,0 +1,19 @@
+class AddTreatmentGroupMembershipConstraint < ActiveRecord::Migration[5.2]
+  def up
+    patients_in_multiple_experiments =
+      Experimentation::TreatmentGroupMembership
+        .group(:patient_id, :experiment_id)
+        .having("count(*) > 1")
+        .select(:patient_id)
+
+    Experimentation::TreatmentGroupMembership
+      .where(patient_id: patients_in_multiple_experiments)
+      .destroy_all
```

Comment:
> I would suggest exporting and storing as csv on s3 or something as a one-time thing in production, and then truncating the table

Not all the data is invalid, we don't want to delete all rows. The intention of getting rid of these is only to allow the constraint. Deleting these doesn't leave any partial `valid` data that reference these rows either. The state of our schema after these deletions will still be valid, is that's the concern.

Reply:
Is this manner of deletion _intended_ in making the current experiment data _valid_? If not, we should _keep_ the invalid data for analysis so we know exactly what part of it is invalid. For example, we might want to remove one of the memberships instead of deleting from both groups.

We shouldn't _remove_ data just for schema validity without keeping a backup. It's a destructive action, and there's no going back.

My suggestion was to just back up the whole thing and start from scratch for simplicity.

Reply:
>  Is this manner of deletion intended in making the current experiment data valid?

Yes, that is the primary intention.

> we should keep the invalid data for analysis so we know exactly what part of it is invalid. For example, we might want to remove one of the memberships instead of deleting from both groups.

We've already analysed the data (at least this portion of it) while reporting the last experiment. We concluded that there is no way to decide which membership to keep and all of them are equally corrupt.

> We shouldn't remove data just for schema validity without keeping a backup.

This is not worth the effort here imo. We can upload these rows as a CSV to S3 in case we need to look them up in the future.

> My suggestion was to just back up the whole thing and start from scratch for simplicity.

We still might need to lookup the last experiment's results for comparisons/benchmarking, hence trying to avoid this.

Reply:
> We can upload these rows as a CSV to S3 in case we need to look them up in the future.
> We still might need to lookup the last experiment's results for comparisons/benchmarking, hence trying to avoid this.

👍 makes sense. Uploading CSV to S3 is what I meant by "backup" as well. This is good enough I think.

> Yes, that is the primary intention.

Got it. I think I took this too literally :) 
> The intention of getting rid of these is only to allow the constraint.

Reply:
Saved these rows to https://simple-database-backups.s3.ap-south-1.amazonaws.com/treatment_group_membership_export_1Nov.json